### PR TITLE
[To new_cluster] Construct basic framework of ConfigNode in new_cluster

### PR DESCRIPTION
**Pull request by CRZbulabula**
Changes:

- Add confignode package
- Add thrift-confignode
- Add DeviceGroupHashExecutor interface

### CODE REVIEW

**Review by wangchao316 — APPROVED** [review]
LGTM

**Review by OneSizeFitsQuorum — APPROVED** [review]
LGTM

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  private DeviceGroupHashExecutor hashExecutor;
+
+  private Lock partitionTableLock;
+  private PartitionTable partitionTable;
+
+  private LoadBalancer loadBalancer;
+
+  public ConfigManager(String hashExecutorClass, int deviceGroupCount) {
+    setHashExecutor(hashExecutorClass, deviceGroupCount);
+  }
```
Comment (neuyilan) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java:52]:
```suggestion
  public ConfigManager(String hashExecutorClass, int deviceGroupCount) {
    setHashExecutor(hashExecutorClass, deviceGroupCount);
    this.partitionTableLock = new ReentrantLock();
    this.partitionTable = new PartitionTable();
    this.loadBalancer = new LoadBalancer(partitionTableLock, partitionTable);
  }
```

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  public ConfigManager() {
+    ConfigNodeConf config = ConfigNodeDescriptor.getInstance().getConf();
+
+    setHashExecutor(config.getDeviceGroupHashExecutorClass(), config.getDeviceGroupCount());
+
+    this.partitionTableLock = new ReentrantLock();
+    this.partitionTable = new PartitionTable();
+
+    this.loadBalancer = new LoadBalancer(partitionTableLock, partitionTable);
+  }
```
Comment (neuyilan) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java:63]:
```suggestion
  public ConfigManager() {
    ConfigNodeConf config = ConfigNodeDescriptor.getInstance().getConf();
    this(config.getDeviceGroupHashExecutorClass(), config.getDeviceGroupCount())
  }
```

File: confignode/src/main/java/org/apache/iotdb/confignode/partition/DataPartitionRule.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * DataPartitionRule is used to hold real-time write-load allocation rules i.e. rules = [(0, 0.3),
+ * (1, 0.2), (2. 0.5)] means allocate 30% of the write-load to VSGGroup-0 and 20% to vsgGroup-1 and
+ * 50% to VSGGroup-2
+ */
+public class DataPartitionRule {
+  // List<Pair<VSGGroupID, Write allocation ratio>>
```
Comment (neuyilan) [inline confignode/src/main/java/org/apache/iotdb/confignode/partition/DataPartitionRule.java:32]: As we have changed VSG concept to region, it is better to change the comment also.

File: thrift-confignode/src/main/thrift/confignode.thrift [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+  rpc.TSStatus deleteStorageGroup(1:i64 sessionId, 2:string storageGroup);
+
+  rpc.TSStatus deleteStorageGroups(1:i64 sessionId, 2:list<string> storageGroups);
+
+  i32 getSchemaPartition(1:i64 sessionId, 2:string device)
+
+  list<i32> getDataPartition(1:i64 sessionId, 2:string device, 3:list<i64> times)
+
+  list<i32> getLatestDataPartition(1:i64 sessionId, 2:string device)
```
Comment (neuyilan) [inline thrift-confignode/src/main/thrift/confignode.thrift:35]: What is the meaning of the Latest?

**Review by neuyilan — COMMENTED** [review]
LGTM, only minor issues, PTAL

File: confignode/src/main/java/org/apache/iotdb/confignode/service/thrift/impl/ConfigServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+import org.apache.iotdb.confignode.manager.ConfigManager;
+import org.apache.iotdb.confignode.rpc.thrift.ConfigIService;
+import org.apache.iotdb.service.rpc.thrift.TSStatus;
+
+import org.apache.thrift.TException;
+
+import java.util.List;
+
+/** ConfigServiceImpl exposes the interface that interacts with the DataNode */
+public class ConfigServiceImpl implements ConfigIService.Iface {
```
Comment (qiaojialin) [inline confignode/src/main/java/org/apache/iotdb/confignode/service/thrift/impl/ConfigServiceImpl.java:30]:
```suggestion
public class ConfigNodeRPCServer implements ConfigIService.Iface {
```

File: confignode/src/main/java/org/apache/iotdb/confignode/service/thrift/impl/ConfigServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+import org.apache.iotdb.confignode.manager.ConfigManager;
+import org.apache.iotdb.confignode.rpc.thrift.ConfigIService;
+import org.apache.iotdb.service.rpc.thrift.TSStatus;
+
+import org.apache.thrift.TException;
+
+import java.util.List;
+
+/** ConfigServiceImpl exposes the interface that interacts with the DataNode */
```
Comment (qiaojialin) [inline confignode/src/main/java/org/apache/iotdb/confignode/service/thrift/impl/ConfigServiceImpl.java:29]:
```suggestion
/** ConfigNodeRPCServer exposes the interface that interacts with the DataNode */
```

File: confignode/src/test/resources/iotdb-confignode.properties [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+####################
+
+# Number of DeviceGroups per StorageGroup
+# Datatype: int
+# device_group_count=10000
+
+# DeviceGroup hash algorithm
+# Datatype: string
+# Currently, BKDR, AP, JS or SDBM hash algorithm is available
+# device_group_hash_algorithm=BKDR
```
Comment (qiaojialin) [inline confignode/src/test/resources/iotdb-confignode.properties:31]: Do you use this in tests? Keep consistent with the properties with the main resource.(full class name)

File: thrift-confignode/src/main/thrift/confignode.thrift (hunk 4 above)
Comment (qiaojialin) [inline thrift-confignode/src/main/thrift/confignode.thrift:35]: what's this for?

File: thrift-confignode/src/main/thrift/confignode.thrift [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+service ConfigIService {
+  rpc.TSStatus setStorageGroup(1:i64 sessionId, 2:string storageGroup);
+
+  rpc.TSStatus deleteStorageGroup(1:i64 sessionId, 2:string storageGroup);
+
+  rpc.TSStatus deleteStorageGroups(1:i64 sessionId, 2:list<string> storageGroups);
+
+  i32 getSchemaPartition(1:i64 sessionId, 2:string device)
+
+  list<i32> getDataPartition(1:i64 sessionId, 2:string device, 3:list<i64> times)
```
Comment (neuyilan) [inline thrift-confignode/src/main/thrift/confignode.thrift:33]: It is better to add one interface which can get a multi device's data partition.

**Review by neuyilan — APPROVED** [review]

File: thrift-confignode/src/main/thrift/confignode.thrift (hunk 8 above)
Comment (qiaojialin) [inline thrift-confignode/src/main/thrift/confignode.thrift:33]: (1) make the times meaning clear: is this the data timestamp or partition start time? If data timestamp, we need to calculate the start time, if already receives a partition start time, we could eliminate one calculation.
(2) need to add two interfaces
  list<i32> getDataPartition(1:i64 sessionId, 2:string deviceGroupId, 3:list<i64> partitionStartTime)
  list<i32> getDataPartition(1:i64 sessionId, 2:list<string> deviceGroupId, 3:i64 partitionStartTime)

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java (hunk 1 above)
Comment (CRZbulabula) [reply confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java:52]: This is a test only constructor. There is no need to create PartitionTable and LoadBalancer for now

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java (hunk 2 above)
Comment (CRZbulabula) [reply confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java:63]: This constructor may be changed latter.  Maybe there is not need to change the code structure for a while.

File: thrift-confignode/src/main/thrift/confignode.thrift (hunk 4 above)
Comment (CRZbulabula) [reply thrift-confignode/src/main/thrift/confignode.thrift:35]: "Latest" means the currently dataPartition on time.

File: confignode/src/assembly/resources/conf/iotdb-confignode.properties [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+# Datatype: int
+# device_group_count=10000
+
+# DeviceGroup hash algorithm
+# Datatype: String
+# These hashing algorithms are currently supported:
+# 1. org.apache.iotdb.confignode.manager.hash.BKDRHashExecutor(Default)
+# 2. org.apache.iotdb.confignode.manager.hash.APHashExecutor
+# 3. org.apache.iotdb.confignode.manager.hash.JSHashExecutor
+# 4. org.apache.iotdb.confignode.manager.hash.SDBMHashExecutor
```
[the remark this answers is not in the document]
Comment (CRZbulabula) [reply confignode/src/assembly/resources/conf/iotdb-confignode.properties:34]: According to the test results so far, different hashing algorithms perform similarly under random tests. Maybe later in UserGuide?

File: thrift-confignode/src/main/thrift/confignode.thrift (hunk 4 above)
Comment (CRZbulabula) [reply thrift-confignode/src/main/thrift/confignode.thrift:35]: This is a redundant interface, I will delete it.

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+  private void setHashExecutor(String hashExecutorClass, int deviceGroupCount) {
```
[the remark this answers is not in the document]
Comment (CRZbulabula) [reply confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java:65]: After discussion with 思屹, he thinks DataNode also needs to get the hash algorithm for the device group. I think maybe a new PR is needed to create a common tool package between DataNode and ConfigNode. And in this new PR I will solve this problem.

File: thrift-confignode/src/main/thrift/confignode.thrift (hunk 8 above)
Comment (CRZbulabula) [reply thrift-confignode/src/main/thrift/confignode.thrift:33]: This is the older interface. The new interface is defined in the following documentation:

- [分区接口定义](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=199536930)
- [ConfigNode thrift-rpc 接口定义](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=199538743)

File: confignode/src/assembly/resources/conf/iotdb-confignode.properties [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+# under the License.
+#
+
+####################
+### DeviceGroup Configuration
+####################
+
+# Number of DeviceGroups per StorageGroup
+# Datatype: int
+# device_group_count=10000
```
[the remark this answers is not in the document]
Comment (CRZbulabula) [reply confignode/src/assembly/resources/conf/iotdb-confignode.properties:26]: The DeviceGroup hash algorithm also needs to be able be used by DataNode, so I think maybe a new PR is needed to create a common tool package between ConfigNode and DataNode. I will solve this problem in the new PR.

File: thrift-confignode/src/main/thrift/confignode.thrift (hunk 4 above)
Comment (CRZbulabula) [reply thrift-confignode/src/main/thrift/confignode.thrift:35]: This is the older interface. The new interface is defined in the following documentation:

- [分区接口定义](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=199536930)
- [ConfigNode thrift-rpc 接口定义](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=199538743)

File: confignode/src/assembly/resources/conf/iotdb-confignode.properties (hunk 9 above)
[the remark this answers is not in the document]
Comment (qiaojialin) [reply confignode/src/assembly/resources/conf/iotdb-confignode.properties:34]: ok

**Review by qiaojialin — APPROVED** [review]

**Pull request merged**